[WIP][JIT] Add ScriptModule._reconstruct

SplitInfinity opened the pull request:
**Summary**
This commit adds an instance method `_reconstruct` that permits users
to reconstruct a `ScriptModule` from a given C++ `Module` instance.

**Testing**
This commit adds a unit test for `_reconstruct`.

**Fixes**
This pull request fixes #33912.

SplitInfinity commented on test/test_jit_py3.py:
```diff
@@ -469,6 +469,53 @@ def foo():
+
+        inp = torch.rand((3))
+
+        # Check that both modules produce the same output.
+        with torch.no_grad():
+            m_out = m(inp)
+            n_out = n(inp)
+            self.assertEqual(m_out, n_out)
+
+        # Check that ignored method is still intact.
+        self.assertEqual(inp, n.ignored_method(inp))
```
I'd like to do more equality checks here but I'm not sure how.

jerryzh168 replied:
we want something easier, we are calling reconstruct with the same object, but with a changed `cpp_module`

SplitInfinity commented:
I think this is what @jerryzh168 was looking for but I think it's very specific to their use case and not generalizable because there is no way to figure out what `ignored` methods there were on the original `nn.Module` that was used to build the C++ module that is passed to `_reconstruct`.

There are some attributes that could also have existed on the original `nn.Module` that were copied to the corresponding `ScriptModule` by marking them with `@_copy_to_script_wrapper`. These are also not recoverable from the C++ module.

jerryzh168 commented on torch/jit/__init__.py:
```diff
@@ -1743,6 +1743,32 @@ def _construct(cpp_module, init_fn):
+            Arguments:
+                cpp_module: The C++ module that this RecursiveScriptModule will be rebuilt around.
+            """
+            self.__init__(cpp_module)
```
do you need to delete existing attributes in order to restore the `self` to the state before `__init__`?

SplitInfinity replied:
I think the only attributes on the `ScriptModule` are:
1. memoized attributes for function calls to the underlying C++ module
2. ignored functions that run in Python 
3. other attributes marked for copying from the C++ to `ScriptModule` (internal API not used often)
4. attributes inherited from `nn.Module`

As discussed before, #2 and #3 are not recoverable from the C++ module (not a problem for your use case). There is code below to clear out all `ScriptFunctions` from `self.__dict__` which should take care of #3. 

@suo pls confirm.

suo replied:
I think you are correct. `ScriptModule` is a Python wrapper that provides an `nn.Module`-like view over the C++ ScriptModule, so `reconstruct` should work by just swapping out the cpp module and updating the viewing machinery.

The place where this breaks down is places where, for debug purposes, `ScriptModule` is not a pure view over `cpp_module`, like `@ignore`. It might be good to error if we detect this cases.

SplitInfinity replied:
If I was writing this feature in isolation I would agree, but it looks like @jerryzh168 wants to keep those around? In his narrow use case, it looks like the pre- and post-reconstruction types are basically the same type so these `@ignore`d methods apply to the post-reconstruction type too?

jerryzh168 replied:
yeah, if it does not fit the correct behavior of what _reconstruct is supposed to do, maybe we can give it a more specific name, like `_reconstruct_same_instance`.

jerryzh168 reviewed (approved):
LGTM, thanks! @suo please take look as well.

jerryzh168 commented on test/test_jit_py3.py:
```diff
@@ -486,6 +486,53 @@ def foo():
+        m = torch.jit.script(M(3))
+        n = torch.jit.script(N())
+
+        n._reconstruct(m._c)
```
could you compare the result of `_reconstruct` and `wrap_cpp_module` on `m`? I don't think we need to compare `m` and `n`.

SplitInfinity commented:
ping @suo

jerryzh168 commented:
@SplitInfinity there is a conflict with master btw
> Conflicting files
> test/test_jit_py3.py

suo reviewed (approved):
I think

Commit:
[JIT] Add ScriptModule._reconstruct

Summary:
This commit adds an instance method _reconstruct that permits users
to reconstruct a ScriptModule from a given C++ Module instance.

Test Plan:
This commit adds a unit test for _reconstruct.

Fixes:
This commit fixes #33912.

jerryzh168 commented on test/test_jit_py3.py:
```diff
@@ -486,6 +486,53 @@ def foo():
+        n = torch.jit.script(N())
+
+        n._reconstruct(m._c)
+
```
could you check n.graph == m.graph?
